### PR TITLE
[READY] Reworks the Venus Human Trap

### PR DESCRIPTION
**Pull request by IndieanaJones**
## About The Pull Request

Fixes the currently broken implementation of the venus human trap, a mob spawned by the flowering mutation of kudzu.  They're supposed to attach vines to you and reel you in (I think), but that functionality was entirely busted, only meleeing you with their vines.  Along with this, sentient venus human traps could attack you as fast as they could click, pretty much killing anyone instantly.  So, I completely wiped them and started from scratch, ending up with the following:

Venus Human Trap
- Can attach vines to their targets within a 5-tile radius.
- Vines stun carbons for 2 seconds when first applied, and will attempt to bring the affected closer to the plant.  If the target gets far enough away from the plant or shares the same tile as the plant, the vine will be destroyed.
- If a plant tries to attach a vine to something they already have a vine on, they will reel everything they can towards themselves.  This also happens passively every second.
- Plants heal 10% of their max health when attacking with melee if their target is alive.
- Sentient plants can attach their vines to a wide variety of things, allowing them to take advantage of the damage dealt when walking though them.
- Plants can be possessed by ghosts akin to spiders born from a sentient mother.

Along with all that, I also autodoc'd the whole file as well, as well as fixing the apparently broken feature of kudzu not affecting plant-based creatures (venus human traps and podpeople).

## Why It's Good For The Game

Fixes an entirely broken mechanic to be functional, along with adding a new role for ghosts to possibly take up.

## Changelog
:cl: Indie-ana Jones
balance: Venus Human Traps have mutated to better hunt their prey, becoming smarter and able to attach vines to their prey.
fix: Plant-based creatures being affected by kudzu has been fixed.
/:cl:

### CODE REVIEW

**Comment by carshalash** [comment]
Needs explosive immunity for aggressive growing kudzu.

**Comment by ATH1909** [comment]
Oooooh, the sentience thing looks like it'll make kudzu a LOT more viable to spread as an antag botanist. I like it.

Are you sure you can't let plants pull objects or unanchored structures towards themselves with their vines? I feel like having it only pull mobs is kind of arbitrary.

**Comment by Pumpkinoe** [comment]
can they speak?

**Comment by IndieanaJones** [comment]
> can they speak?

As of right now, yes.

**Comment by trollbreeder** [comment]
Make them only able to speak plant language, like how slimes can only speak slime.

**Comment by Pumpkinoe** [comment]
make sure they get a message that tells them "yo don't harm pod people or other plants pls" if they're sentient

**Comment by IndieanaJones** [comment]
> 
> 
> make sure they get a message that tells them "yo don't harm pod people or other plants pls" if they're sentient

They get a message on what they should do when they spawn in.

**Comment by Pumpkinoe** [comment]
from what I see it just tells them feast on those who oppose them
what if a pod person opposes them? O_O

**Comment by IndieanaJones** [comment]
> 
> 
> from what I see it just tells them feast on those who oppose them
> what if a pod person opposes them? O_O

Fair game, really.  If the pod person goes against the kudzu, they're opposition.

**Comment by Pumpkinoe** [comment]
fair enough then
does that mean that they can be cool with some human that's giving them meat?

File: code/modules/events/spacevine.dm [hunk 1]
```diff
@@ -251,7 +251,7 @@
 	holder.obj_integrity = holder.max_integrity
 
 /datum/spacevine_mutation/woodening/on_hit(obj/structure/spacevine/holder, mob/living/hitter, obj/item/I, expected_damage)
-	if(I && I.get_sharpness())
+	if(I.get_sharpness())
```
Comment (spookydonut) [inline code/modules/events/spacevine.dm]:
```suggestion
	if(I?.get_sharpness())
```
`I` can be null

File: code/modules/events/spacevine.dm [hunk 2]
```diff
@@ -540,8 +540,8 @@
 		return TRUE
 
 /proc/isvineimmune(atom/A)
-	. = FALSE
 	if(isliving(A))
 		var/mob/living/M = A
 		if(("vines" in M.faction) || ("plants" in M.faction))
-			. = TRUE
+			return TRUE
```
Comment (spookydonut) [inline code/modules/events/spacevine.dm]: weird flex but this is just code churn to do exactly the same thing

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 3]
```diff
@@ -25,35 +34,51 @@
 		B.sleep_time = 10 //these shouldn't move, so let's slow down updates to 1 second (any slower and the deletion of the vines would be too slow)
 	addtimer(CALLBACK(src, .proc/bear_fruit), growth_time)
 
+/**
+  * Spawns a venus human trap, then qdels itself.
+  *
+  * Displays a message, spawns a human venus trap, then qdels itself.
+  */	
+
```
Comment (spookydonut) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
```

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 4]
```diff
@@ -25,35 +34,51 @@
...
+  *
+  * The result of a kudzu flower bud, these enemies use vines to drag prey close to them for attack.
+  *
+  * A carnivorious plant which uses vines to catch and ensnare prey.  Spawns from kudzu flower buds.
+  * Each one has a maximum of four vines, which can be attached to a variety of things.  Carbons are stunned when a vine is attached to them, and movable entities are pulled closer over time.
+  * Attempting to attach a vine to something with a vine already attached to it will pull all movable targets closer on command.
+  * Once the prey is in melee range, melee attacks from the venus human trap heals itself for 10% of its max health, assuming the target is alive.
+  * Akin to certain spiders, venus human traps can also be possessed and controlled by ghosts.
+  *
+  */
```
Comment (spookydonut) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
```

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 5]
```diff
@@ -62,75 +87,96 @@
...
+
+/**
+  * Sets a ghost to control the plant if the plant is eligible
+  *
+  * Asks the interacting ghost if they would like to control the plant.
+  * If they answer yes, and another ghost hasn't taken control, sets the ghost to control the plant.
+  * Arguments:
+  * * mob/user - The ghost to possibly control the plant
+  */
+
```
Comment (spookydonut) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
```

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 6]
```diff
@@ -62,75 +87,96 @@
...
+	log_game("[key_name(src)] took control of [name].")
+
+/**
+  * Manages how the vines should affect the things they're attached to.
+  *
+  * Pulls all movable targets of the vines closer to the plant
+  * If the target is on the same tile as the plant, destroy the vine
+  * Removes any QDELETED vines from the vines list.
+  */
+
```
Comment (spookydonut) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
```

**Review by spookydonut — CHANGES_REQUESTED** [review]

**Comment by fludd12** [comment]
_Does it have to be human?_

**Comment by fludd12** [comment]
> _Does it have to be human?_

**_DOES IT HAVE TO BE MINE?!_**

**Comment by Pumpkinoe** [comment]
Fix aggressive vines killing venus man traps btw

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 7]
```diff
@@ -1,4 +1,13 @@
...
+/**
+  * Kudzu Flower Bud
+  *
+  * A flower created by flowering kudzu which spawns a venus human trap after a certain amount of time has passed.
+  *
+  * A flower created by kudzu with the flowering mutation.  Spawns a venus human trap after 2 minutes under normal circumstances.
+  * Also spawns 4 vines going out in diagonal directions from the bud.  Any living creature not aligned with plants is damaged by these vines.
+  * Once it grows a venus human trap, the bud itself will destroy itself.
+  *
+  */
```
Comment (optimumtact) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]: NO SPACE BEFORE THE TYPE!!!

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 8]
```diff
@@ -62,75 +84,94 @@
...
+	/// A list of all the plant's vines
+	var/list/vines = list()
+	/// The maximum amount of vines a plant can have at one time
+	var/max_vines = 4
+	/// How far away a plant can attach a vine to something
+	var/vine_grab_distance = 5
+	/// Whether or not this plant is ghost possessable
+	var/playable_plant = TRUE
+
+/mob/living/simple_animal/hostile/venus_human_trap/Life(mapload)
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
/mob/living/simple_animal/hostile/venus_human_trap/Life()
```

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 9]
```diff
@@ -62,75 +84,94 @@
...
+	if(isliving(target))
+		var/mob/living/L = target
+		if(L.stat != DEAD)
+			adjustHealth(-maxHealth * 0.1)
 
 /mob/living/simple_animal/hostile/venus_human_trap/OpenFire(atom/the_target)
+	for(var/datum/beam/B in vines)
+		if(B.target == the_target)
+			pull_vines()
+			ranged_cooldown = world.time + (ranged_cooldown_time / 2)
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
			ranged_cooldown = world.time + (ranged_cooldown_time * 0.5)
```
Prefer the use of multiplication over division.

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 10]
```diff
@@ -62,75 +84,94 @@
...
-	Beam(the_target, "vine", time=dist*2, maxdistance=dist+2, beam_type=/obj/effect/ebeam/vine)
-	the_target.attack_animal(src)
+	vines += Beam(the_target, "vine", time=INFINITY, maxdistance = vine_grab_distance, beam_type=/obj/effect/ebeam/vine)
+	if(isliving(the_target))
+		var/mob/living/L = the_target
+		L.Paralyze(20)
+	ranged_cooldown = world.time + ranged_cooldown_time
 
+/mob/living/simple_animal/hostile/venus_human_trap/Login()
+	..()
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
	. = ..()
```

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 11]
```diff
@@ -62,75 +84,94 @@
...
+  * Asks the interacting ghost if they would like to control the plant.
+  * If they answer yes, and another ghost hasn't taken control, sets the ghost to control the plant.
+  * Arguments:
+  * * mob/user - The ghost to possibly control the plant
+  */
+/mob/living/simple_animal/hostile/venus_human_trap/proc/humanize_plant(mob/user)
+	if(key || !playable_plant || stat)
+		return
+	var/plant_ask = alert("Become a venus human trap?", "Are you reverse vegan?", "Yes", "No")
+	if(plant_ask == "No" || !src || QDELETED(src))
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
	if(plant_ask == "No" || QDELETED(src))
```

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 12]
```diff
@@ -62,75 +84,94 @@
...
+/mob/living/simple_animal/hostile/venus_human_trap/proc/pull_vines()
+	for(var/datum/beam/B in vines)
+		if(istype(B.target, /atom/movable))
+			var/atom/movable/AM = B.target
+			if(!AM.anchored)
+				step(AM,get_dir(AM,src))
+		if(get_dist(src,B.target) == 0)
+			B.End()
+		if(QDELETED(B))
+			vines -= B
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]: Handle this in the beam's `Destroy()` or better yet register for a destruction signal.

Comment (IndieanaJones) [reply code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]: Could you explain what you mean by a destruction signal?

Comment (Rohesie) [reply code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]: Do you know how signals work? Poke me on #coding-general if you want it explained. The signal we'd be listening to would be `COMSIG_PARENT_QDELETING` in this case.

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 13]
```diff
@@ -173,5 +176,14 @@
...
+
+/**
+  * Removes a vine from the list.
+  *
+  * Removes the vine from our list.
+  * Called specifically when the vine is about to be destroyed, so we don't have any null references.
+  * Arguments:
+  * * datum/beam/vine - The vine to be removed from the list.
+  */
+mob/living/simple_animal/hostile/venus_human_trap/proc/remove_vine(datum/beam/vine)
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]:
```suggestion
mob/living/simple_animal/hostile/venus_human_trap/proc/remove_vine(datum/beam/vine, force)
```
The signal sends two args: `SEND_SIGNAL(D, COMSIG_PARENT_QDELETING, force)`
The first one is always the source, so you have it right, the second one here is `force`. Just for reference.

File: code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm [hunk 14]
```diff
@@ -122,7 +122,10 @@
 		for(var/obj/O in T)
 			if(O.density)
 				return
-	vines += Beam(the_target, "vine", time=INFINITY, maxdistance = vine_grab_distance, beam_type=/obj/effect/ebeam/vine)
+	
+	var/datum/beam/newVine = Beam(the_target, "vine", time=INFINITY, maxdistance = vine_grab_distance, beam_type=/obj/effect/ebeam/vine)
+	RegisterSignal(newVine, COMSIG_PARENT_QDELETING, .proc/remove_vine, newVine)
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/venus_human_trap.dm]: Maybe something like `.proc/remove_from_vines`? Feel free to disregard the suggestion if you feel `remove_vine` is more clear.

**Comment by Rohesie** [comment]
Test your changes after you are done and comment to let us know if it's ready.

**Comment by IndieanaJones** [comment]
Just ran some tests, everything seems to be working just fine.

**Pull request merged**